### DISCUSSION TITLE
Articoli (items_list) - Inviare Attributi avanzati via API in fatture elettroniche alle PA

### DISCUSSION THREAD

**Discussion by mbargigia** [fattureincloud/api#196] [Q&A] [answered]
Salve, emettiamo fatture elettroniche alle PA (in particolare al Ministero M.I.U.R. di Roma) e, da qualche anno, abbiamo sempre avuto il problema di non poter inviare in automatico il più importante dei dati da includere nelle fatture.

Questi dati sono i Buoni MIUR che riceviamo dai Docenti e per i quali fatturiamo il Ministero.
I codici dei buoni sono stringhe di 7 caratteri alfanumerici (case-sensitive) e vanno nel campo "Codice" di uno specifico "Attributo avanzato". Anche il valore del Buono deve venire inserito nel campo "valore" apposito.
Continuiamo ad inserirli manualmente ma, visto il numero sempre crescente di Docenti che utilizzano detti buoni per i nostri servizi, la procedura sta diventando insopportabile.

In pratica, dopo aver predisposto nel nostro sistema interno la fattura PA (nel ns. sistema i dati suddetti sono già tutti disponibili. Basterebbe inviarli, ma non si può fare) e dopo averla caricata tramite API in FattureInCloud, dobbiamo ogni volta:
a) entrare nel vostro sito;
b) modificare una ad una le fatture inserendo A MANO, **per ogni singolo articolo di ogni fattura**, i dati mancanti nel campo (Attributi Avanzati) 2.2.1.3.1 (n.buono) e 2.2.1.3.2 (valore buono) e 
c) salvare la fattura.
Se non facciamo questo, le fatture elettroniche vengono rifiutate dal M.I.U.R: il Ministero ci ha confermato che i dati DEBBONO apparire in quei campi e da nessun'altra parte, poiché la fattura viene vagliata e convalidata elettronicamente.

La domanda quindi è.... quando mai diventerà possibile passare in automatico questi dati, **che sono a tutti gli effetti parte integrante della fattura PA ma che le API non prevedono?**

Un esempio JSON (API v.2) di come si potrebbe implementare ed inviare, **per ogni item in items_list**, l'invio degli attributi avanzati: 
{ "advanced_attributes" : { "2.2.1.3" : { "code" : "AxjMnop" , "value" : 260 } } }
oppure
{ "advanced_attributes" : [ { "2.2.1.3.1" : "AxjMnop" } , { "2.2.1.3.2" : 260 } ] }

Idealmente, la stessa cosa dovrebbe poter essere fatta anche per tutti gli altri Attributi, cambiando il codice che li identifica.

Sin dall'inizio, abbiamo "digerito" la manualità della cosa, seppur sgradita, ma oggi capita di dover dedicare un dipendente per ore a completare le fatture nel vostro sito. Il lavoro è tedioso (è facile sbagliare a scrivere i codici dei buoni. E se si sbaglia, passano mesi prima di poter ottenere i pagamenti), alquanto lungo e, dopotutto, non è giustificabile in un servizio di fatture elettroniche completo.

Gradirei un vostro commento al riguardo. Meglio ancora... gradirei sapere che diventerà possibile farlo :-)

Grazie e Buon Natale/Buon anno a tutti

<img width="795" alt="image002" src="https://user-images.githubusercontent.com/61627937/207964913-d81c75af-9ee0-4023-b032-100d4dbd90f6.png">
<img width="584" alt="image001" src="https://user-images.githubusercontent.com/61627937/207964924-b0bf199a-b925-4c5f-980b-a222c90ed86c.png">

**Comment by valmoz** [comment, marked as answer]
Buongiorno @mbargigia 
(quasi) tutto ciò che lei può fare attraverso l'interfaccia grafica di Fatture in Cloud può essere fatto anche attraverso le API, in quanto il nostro sito web utilizza le stesse API. Quindi si tratta solamente di trovare l'informazione riguardante il come applicarlo lato API.

In questo caso, la guida che fa al caso suo è la [E-Invoice XML Customization Guide](https://developers.fattureincloud.it/docs/guides/e-invoice-xml-customisation#-items-list).

In particolare, dovrebbe bastare utilizzare il campo ei_raw nel modo seguente:
```json
 "ei_raw":{
    "DettaglioLinee":{
       "CodiceArticolo":[
          {
             "CodiceTipo":"AxjMnop",
             "CodiceValore": "260"
          }
       ]
    }
 }
```
Spero di non aver sbagliato a leggere lo schema ufficiale della PA, ma dovrebbe essere corretto :D

Faccia attenzione a due cose:

- CodiceArticolo deve essere un array JSON, in quanto l'XML risultante può contenere più elementi CodiceArticolo
- Anche i valori numerici devono essere passati in formato stringa, in questo caso "260"

Ci faccia sapere se questo risolve le sue necessità e se ha ulteriori domande a riguardo.

Grazie
Mauro

**Reply by mbargigia** [reply]
Grazie Mauro!
Oggi a Milano è una giornata uggiosa. Ma appena ho letto la sua risposta è come se fosse uscito il sole!
Mi pareva strano che non si potesse fare, ma per due volte in passato (con API v.1) mi era stato detto di no (o non mi aveva risposto nessuno esaurientemente come ha fatto lei). 
Non ha idea di quanto tempo ci farà risparmiare questa notizia.
Grazie ancora e saluti

**Reply by valmoz** [reply]
Bene, sono felice di esserle stato utile!
è proprio per questo che stiamo migrando verso le API v2, sono molto più complete rispetto alle v1.

Buone feste
Mauro
